### PR TITLE
fix cat

### PR DESCRIPTION
**Pull request by j-ogas**
WIP.

Addresses #121.

TODO:
- [x] update test target to generic file without user specific data
- [x] make travis happy

### CODE REVIEW

File: test/tests/cat_tests.py [hunk 1]
```diff
@@ -26,10 +26,22 @@ def test_cat(self):
...
+#!/bin/bash
+
+# The first (and only) argument of the build script is the test id.
+export TEST_ID={}
+export PAV_CONFIG_FILE={}
+source {}
+
+# Perform the sequence of test commands.
+echo "Hello World."
+""".format(test.id, test.TODO, test.TODO)
```
Comment (j-ogas) [inline test/tests/cat_tests.py]: @Paul-Ferrell Where can I get the format field values from the quick test object?

File: test/tests/cat_tests.py [hunk 2]
```diff
@@ -26,10 +26,22 @@ def test_cat(self):
...
 
         for arg_set in arg_sets:
             args = arg_parser.parse_args(arg_set)
             cat_cmd.run(self.pav_cfg, args)
-            output.fprint(cat_cmd.outfile)
-            output.fprint(cat_cmd.errfile)
+            # FIXME: always true
+            if cat_cmd.outfile != true_out:
+                return errno.ENOMSG
```
Comment (j-ogas) [inline test/tests/cat_tests.py]: Test always passes with this line.

Comment (Paul-Ferrell) [reply test/tests/cat_tests.py]: cat_cmd.outfile is a file-like object. You have to read it. `cat_cmd.outfile.read()`

File: test/tests/cat_tests.py [hunk 3]
```diff
@@ -26,10 +26,22 @@ def test_cat(self):
         cat_cmd.errfile = io.StringIO()
 
         arg_parser = arguments.get_parser()
-        arg_sets = (['cat', str(test.id), 'build.sh'],)
```
Comment (Paul-Ferrell) [inline test/tests/cat_tests.py]: To get a nice long file to mess with: 

```python
test_cfg = self._quick_test_cfg()
test_cfg['run']['cmds'] = ['echo "hello world"'] * 1000
test = self._quick_test(test_cfg)
```

File: test/tests/cat_tests.py (hunk 2 above)
Comment (Paul-Ferrell) [reply test/tests/cat_tests.py]: Use the test object assert statements. 
```python
self.assertEqual(thing1, thing2)
#In this case:
true_out = (test.path/arg_set[-1]).open().read()
self.assertEqual(cat_cmd.outfile.read(), true_out)
```

File: lib/pavilion/plugins/commands/cat.py [hunk 4]
```diff
@@ -54,11 +54,8 @@ def print_file(file):
 
     try:
         with file.open() as file:
-            while True:
```
Comment (Paul-Ferrell) [inline lib/pavilion/plugins/commands/cat.py]: The way you did this before was better; you don't want to read a whole arbitrary file into memory at once.
Just add `end=""` to your print statement. (fprint passes extra keyword arguments to print())

File: lib/pavilion/plugins/commands/cat.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def print_file(file):
                 block = file.read(4096)
                 if not block:
                     break
-                output.fprint(block, file=sys.stdout)
+                print(block, file=sys.stdout, end="")
```
Comment (Paul-Ferrell) [inline lib/pavilion/plugins/commands/cat.py]:
```
output.fprint(block, file=self.outfile, end="")
```

File: test/tests/cat_tests.py [hunk 6]
```diff
@@ -19,17 +19,21 @@ def teatDown(self):
...
         cat_cmd.outfile = io.StringIO()
         cat_cmd.errfile = io.StringIO()
 
         arg_parser = arguments.get_parser()
-        arg_sets = (['cat', str(test.id), 'build.sh'],)
-
+        arg_sets = (['cat', str(test.id), 'run.tmpl'],)
         for arg_set in arg_sets:
             args = arg_parser.parse_args(arg_set)
+            sys.stdout = cat_cmd.outfile
```
Comment (Paul-Ferrell) [inline test/tests/cat_tests.py]: no!!!

File: lib/pavilion/plugins/commands/cat.py [hunk 7]
```diff
@@ -44,28 +45,27 @@ def run(self, pav_cfg, args):
...
+                    output.fprint(block, file=self.outfile, end="")
 
-    try:
-        with file.open() as file:
-            while True:
-                block = file.read(4096)
-                if not block:
-                    break
-                output.fprint(block, file=sys.stdout)
+        except IsADirectoryError:
```
Comment (Paul-Ferrell) [inline lib/pavilion/plugins/commands/cat.py]: You should also catch OSError, IOError, and PermissionError. You can get them as a group, and then give a generic message like:
"Error reading file '{}': {}".format(file_path, err)

File: lib/pavilion/plugins/commands/cat.py [hunk 8]
```diff
@@ -44,28 +45,25 @@ def run(self, pav_cfg, args):
...
-
-    except FileNotFoundError:
-        output.fprint("file '{}' does not exist.".format(file), sys.stderr,
-                      color=output.RED)
-        return errno.EEXIST
+        except (FileNotFoundError,
+                IsADirectoryError,
+                OSError,
+                PermissionError) as err:
+            output.fprint("Error reading file '{}': {}".format(file), err,
```
Comment (Paul-Ferrell) [inline lib/pavilion/plugins/commands/cat.py]: :man_facepalming: - I should be more explicit. For generic errors that you don't know the cause for (OSError, IOError), give a generic message. For errors with an obvious cause, give a _specific_ mesage like you were doing before. 

More specific errors are almost always better. :)

**Pull request merged**